p2p/src/net/libp2p/backend.rs

```diff
@@ -62,57 +49,163 @@ pub struct Backend {
 
     /// Sender for outgoing syncing events
     pub(super) sync_tx: mpsc::Sender<types::SyncingEvent>,
-
-    /// Set of pending connections
-    pub(super) pending_conns: HashMap<PeerId, PendingState>,
-
-    /// Set of established connections
-    pub(super) established_conns: HashSet<PeerId>,
-
-    /// Set of pending requests
-    pub(super) pending_reqs: HashMap<RequestId, ResponseChannel<SyncResponse>>,
-
-    /// Whether mDNS peer events should be relayed to P2P manager
-    pub(super) relay_mdns: bool,
 }
 
 impl Backend {
     pub fn new(
-        swarm: Swarm<types::ComposedBehaviour>,
+        swarm: Swarm<behaviour::Libp2pBehaviour>,
         cmd_rx: mpsc::Receiver<types::Command>,
         conn_tx: mpsc::Sender<types::ConnectivityEvent>,
         gossip_tx: mpsc::Sender<types::PubSubEvent>,
         sync_tx: mpsc::Sender<types::SyncingEvent>,
-        relay_mdns: bool,
     ) -> Self {
         Self {
             swarm,
             cmd_rx,
             conn_tx,
             gossip_tx,
             sync_tx,
-            pending_conns: HashMap::new(),
-            established_conns: HashSet::new(),
-            pending_reqs: HashMap::new(),
-            relay_mdns,
         }
     }
 
-    // TODO: into_fatal()???
-    pub async fn run(&mut self) -> crate::Result<()> {
+    pub async fn on_connection_established(
+        &mut self,
+        peer_id: PeerId,
+        endpoint: ConnectedPoint,
+    ) -> crate::Result<()> {
+        match endpoint {
+            ConnectedPoint::Dialer { .. } => {
+                log::trace!("connection established (dialer), peer id {:?}", peer_id);
+
+                match self.swarm.behaviour_mut().pending_conns.remove(&peer_id) {
```

Comment:
This kind of code style is almost C programming. Everything is mutable and we have a soup of mutations everywhere. Can this be done in such a way where behavior is not mutated by other objects without proper state verification? Can you please rethink the design a little here so that we don't have to do things like this?

Is a method "remove_pending_conns" gonna make me happier? Maybe a little. Depends on whether you just call remove in that function; that won't solve this problem. But basically I'd like a system where all invariants are held during the execution of the program, more than just mutating continuously things like C code does.

Reply:
Yeah this is very much work in progress. The idea is to implement `ConnectionManager` which deals with low-level connectivity events. It implements the `NetworkBehaviour` trait so it perfectly integrates with the changes introduced in this PR. That's the goal but implementing it in this PR would've made this PR even larger than it already is and it would've introduced a lot of new changes which I wanted to avoid in this kind of "moving stuff around" PR.

So in closing, this is bad design and there absolutely is a plan to fix this which integrates with the libp2p `Swarm`/`NetworkBehaviour` design much more nicely. I just haven't started to work on that yet.

Reply:
Alrighty. I'll approve based on that.